topology/probes/istio/virtualservice.go

```diff
@@ -48,3 +50,120 @@ func (h *virtualServiceHandler) Dump(obj interface{}) string {
+}
+
+func getHostsByVirtualService(vs *kiali.VirtualService) []string {
+	hosts := []string{}
+	if vs.Spec["http"] == nil {
+		return hosts
+	}
+	https := vs.Spec["http"].([]interface{})
+	for _, http := range https {
+		httpMap := http.(map[string]interface{})
+		if httpMap["route"] == nil {
+			continue
+		}
+		routes := httpMap["route"].([]interface{})
+		for _, route := range routes {
+			routeMap := route.(map[string]interface{})
+			if routeMap["destination"] == nil {
+				continue
+			}
+			destination := routeMap["destination"].(map[string]interface{})
+			hosts = append(hosts, destination["host"].(string))
+		}
+	}
+	return hosts
+}
+
+func getVirtualServiceByApp(vsCache *k8s.ResourceCache, app string) *kiali.VirtualService {
+	for _, vs := range vsCache.List() {
+		if vs, ok := vs.(*kiali.VirtualService); ok {
+			if vs.Spec["http"] == nil {
+				continue
+			}
+			https := vs.Spec["http"].([]interface{})
+			for _, http := range https {
+				httpMap := http.(map[string]interface{})
+				if httpMap["route"] == nil {
+					continue
+				}
+				routes := httpMap["route"].([]interface{})
+				for _, route := range routes {
+					routeMap := route.(map[string]interface{})
+					if routeMap["destination"] == nil {
+						continue
+					}
+					destination := routeMap["destination"].(map[string]interface{})
+					if destination["host"] == nil {
+						continue
+					}
+					if app == destination["host"].(string) {
+						return vs
+					}
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (vspl *virtualServicePodLinker) GetABLinks(vsNode *graph.Node) (edges []*graph.Edge) {
+	vs := vspl.vsCache.GetByNode(vsNode)
+	if vs, ok := vs.(*kiali.VirtualService); ok {
+		hosts := getHostsByVirtualService(vs)
+		if len(hosts) == 0 {
+			return
+		}
+		pods := k8s.GetPodsByHosts(vspl.podCache, hosts)
+		if len(pods) == 0 {
+			return
+		}
+		for _, pod := range pods {
+			if podNode := vspl.graph.GetNode(graph.Identifier(pod.GetUID())); podNode != nil {
+				id := graph.GenID(string(vsNode.ID), string(podNode.ID), "RelationType", "virtualservice")
+				edges = append(edges, vspl.graph.NewEdge(id, vsNode, podNode, newEdgeMetadata("virtualservice"), ""))
+			}
+		}
+	}
+	return
+}
+
+func (vspl *virtualServicePodLinker) GetBALinks(podNode *graph.Node) (edges []*graph.Edge) {
+	pod := vspl.podCache.GetByNode(podNode)
+	if pod, ok := pod.(*v1.Pod); ok {
+		app := k8s.GetAppByPod(pod)
+		vs := getVirtualServiceByApp(vspl.vsCache, app)
+		if vs == nil {
+			return
+		}
+		if vsNode := vspl.graph.GetNode(graph.Identifier(vs.GetUID())); vsNode != nil {
+			edges = append(edges, vspl.graph.CreateEdge("", vsNode, podNode, newEdgeMetadata("virtualservice"), graph.TimeUTC(), ""))
+		}
+	}
+	return
+}
+
+func newVirtualServicePodLinker(g *graph.Graph, probes map[string]k8s.Subprobe) probe.Probe {
+	vsProbe := probes["virtualservice"]
+	podProbe := k8s.Subprobes["pod"]
+	if vsProbe == nil || podProbe == nil {
+		return nil
+	}
+	return graph.NewResourceLinker(
```

Comment:
maybe create the virtualServicePodLinker first and then you can invoke graph.NewResourceLinker in a single line

Reply:
DONE